d2graph/grid_diagram.go

```diff
@@ -0,0 +1,16 @@
+package d2graph
+
+func (obj *Object) IsGridDiagram() bool {
+	return obj != nil && obj.Attributes != nil &&
+		(obj.Attributes.Rows != nil || obj.Attributes.Columns != nil)
+}
+
+func (obj *Object) ClosestGridDiagram() *Object {
+	if obj.Parent == nil {
+		return nil
+	}
+	if obj.Parent.IsGridDiagram() {
+		return obj.Parent
+	}
+	return obj.Parent.ClosestGridDiagram()
```

Comment:
operate on the obj instead of parent

```
if obj == nil: return nil
if obj.IsGridDiagram(): return obj
else: return obj.Parent.ClosestGridDiagram()
```